antenna/breakpad_resource.py

```diff
@@ -368,45 +393,76 @@ def on_post(self, req, resp):
             resp.body = 'CrashID=%s%s\n' % (self.config('dump_id_prefix'), crash_id)
 
     def hb_run_crashmover(self):
-        """Check to see if it should spawn a crashmover and does if appropriate."""
+        """Spawn a crashmover if there's work to do."""
         # Spawn a new crashmover if there's stuff in the queue and there isn't
         # one currently running
         if self.crashmover_save_queue and self.crashmover_pool.free_count() > 0:
-            self.crashmover_pool.spawn(self.crashmover_process_queue)
+            self.crashmover_pool.spawn(self.crashmover_process_queues)
 
-    def crashmover_process_queue(self):
-        """Process the queue of crashes to save until it's empty.
+    def crashmover_process_queues(self):
+        """Process crashmover work.
 
         Note: This has to be super careful not to lose crash reports. If
-        there's any kind of problem, this must return the crash to the queue.
+        there's any kind of problem, this must return the crash report to the
+        save queue.
 
         """
-        # Process crashes until the queue is empty
-        while self.crashmover_save_queue:
-            crash_report = self.crashmover_save_queue.popleft()
-
-            try:
-                self.crashmover_save(crash_report)
-
-            except Exception:
-                mymetrics.incr('save_crash_exception.count')
-                crash_report.errors += 1
-                logger.exception(
-                    'Exception when processing save queue (%s); error %d/%d',
-                    crash_report.crash_id,
-                    crash_report.errors,
-                    MAX_ATTEMPTS
-                )
-
-                # After MAX_ATTEMPTS, we give up on this crash and move on
-                if crash_report.errors < MAX_ATTEMPTS:
-                    self.crashmover_save_queue.append(crash_report)
-                else:
-                    logger.error('%s: too many errors trying to save; dropped', crash_report.crash_id)
-                    mymetrics.incr('save_crash_dropped.count')
+        while self.crashmover_save_queue or self.crashmover_publish_queue:
+            if self.crashmover_save_queue:
+                crash_report = self.crashmover_save_queue.popleft()
+
+                try:
+                    # Save crash and then toss crash_id in the publish queue
+                    self.crashmover_save(crash_report)
+
+                    crash_report.errors = 0
```

Comment:
Once we've successfully saved the crash report, we zero out the errors and move it to publishing.